### PR TITLE
ENH: Add support for MINC image IO

### PR DESCRIPTION
**Pull request by thewtex**
Issue #4085

This enables MINC Image IO via ITK. This currently supports MINC version 2
(HDF5-based).

### CODE REVIEW

**Comment by vovythevov** [comment]
LGTM ! Thanks for the contribution !

**Comment by thewtex** [comment]
Tested locally on Linux.

CC: @lassoan @jcfr @vfonov @gdevenyi

**Comment by fedorov** [comment]
xref https://github.com/CobraLab/atlases/issues/6

**Comment by thewtex** [comment]
@fedorov Thanks for the note!

I am trying to update ITK's MINC to address the reported Windows build issue...

**Comment by vfonov** [comment]
So, what is the problem now?

**Comment by thewtex** [comment]
ITK's libminc is updated to the latest `develop` branch version here:

  http://review.source.kitware.com/#/c/22329/

This includes: https://github.com/BIC-MNI/libminc/commit/d8cdab3407c130f7b021e11caf355eaf94aee52b but newer Visual Studio may still not work because of https://github.com/BIC-MNI/libminc/issues/77

**Comment by thewtex** [comment]
Spoke too soon: the update does include https://github.com/BIC-MNI/libminc/commit/96793536aee1a7bc9ffcce23f6713e2b75238152

@lassoan could you please test? We should then mark  BIC-MNI/libminc@d8cdab3 as closed if it has been addressed.

**Comment by fedorov** [comment]
@thewtex I can test on mac, if this would help, but I won't find time to rigorously examine the code (I see you assigned me on gerrit), I have no familiarity with this library at all.

Let me know if it would help if I test this branch on mac, I don't want to do if someone else is already did/doing it.

**Comment by thewtex** [comment]
@fedorov Thanks, yes that would help. I only tested on Linux. To test, check out the branch:

  git fetch http://review.source.kitware.com/ITK refs/changes/29/22329/1 && git checkout FETCH_HEAD

and set `Module_ITKIOMINC=1` in ITK's CMake configuration.

To build only `ITKIOMINC` module, 

```
make ITKIOMINC-all
```

To run just the MINC tests:

```
ctest -L ITKIOMINC
```

**Comment by fedorov** [comment]
@thewtex ok I will test ITK as you suggested.

Should we test if this causes any issues with SimpleITK?

Isn't this specific PR premature, as it would need to also include an update to ITK which is not finalized itself yet?

**Comment by thewtex** [comment]
@fedorov thanks!

I do not anticipate any issues with SimpleITK.

Yes, we will need to get the fixes into ITK first, then update Slicer's version of ITK.

**Comment by lassoan** [comment]
@thewtex What should I test? Should I try to build the version in this PR? If a custom ITK version is needed, could you update the ITK repository/git tag in SuperBuild\External_ITKv4.cmake to that version?

**Comment by fedorov** [comment]
Here is the error I am getting on mac compiling ITK branch @thewtex asked me to test:

```
[ 89%] Building C object Modules/ThirdParty/MINC/src/libminc/CMakeFiles/itkminc2.dir/libcommon/minc_error.c.o
In file included from /Users/fedorov/github/ITK/Modules/ThirdParty/MINC/src/libminc/libcommon/minc_error.c:92:
In file included from /Users/fedorov/github/ITK/Modules/ThirdParty/MINC/src/libminc/libsrc/minc_private.h:83:
/Users/fedorov/github/ITK/Modules/ThirdParty/MINC/src/libminc/libsrc/minc.h:171:10: fatal error: 'netcdf.h' file not found
#include <netcdf.h>
         ^
1 error generated.
```

**Comment by lassoan** [comment]
The issue of MINC support came up a couple of times in the past but we failed to add it due that there were always build errors on non-Linux platforms.

Considering persistent build issues, lack of cross-platform interest of MINC developers, and existence of a number of other well-supported alternative file formats, I would consider moving away from this format, maybe even move it from ITK proper to a remote module.

**Comment by vfonov** [comment]
@fedorov - somehow it is configured to search for NetCDF , can you send configuration parameters?

**Comment by fedorov** [comment]
@vfonov: as mentioned earlier by @thewtex, the proposed modification to ITK to include MINC support is in this branch: http://review.source.kitware.com/#/c/22329/. Instructions how to build it are provided here: https://github.com/Slicer/Slicer/pull/716#issuecomment-297834392. If you have a mac, you should be able to reproduce the issue and debug it.

I don't know how and where configuration parameters are passed -- it is configured as a third-party module of ITK, and I don't know the details of implementation.

**Comment by vfonov** [comment]
Here are my updates: http://review.source.kitware.com/#/c/22331/

To test it , I configure with -DITK_USE_SYSTEM_MINC:BOOL=OFF -DModule_ITKIOMINC:BOOL=ON -DModule_ITKIOTransformMINC:BOOL=ON -DModule_ITKMINC:BOOL=ON

To compile: make ITKIOMINC-all ITKIOTransformMINC-all

And to run tests: ctest -L ITKIOTransformMINC -L ITKIOMINC

**Comment by fedorov** [comment]
@vfonov great!

1. ` git fetch http://review.source.kitware.com/ITK refs/changes/31/22331/1  && git checkout FETCH_HEAD`
2. `cd ITK-build && cmake -DITK_USE_SYSTEM_MINC:BOOL=OFF -DModule_ITKIOMINC:BOOL=ON -DModule_ITKIOTransformMINC:BOOL=ON -DModule_ITKMINC:BOOL=ON ../ITK`
3. `make ITKIOMINC-all ITKIOTransformMINC-all`
4. Build worked without problems on mac!
5. All tests passed except ITKIOMINCKWStyleTest, but I think this is due to mis-configuration on my machine.

**Comment by thewtex** [comment]
Hats off to @vfonov for the update!

@lassoan Here is a branch that includes the MINC/ITK update: https://github.com/thewtex/Slicer/tree/minc-io-with-itk-update

**Comment by lassoan** [comment]
@thewtex Thanks, I've started a build of  https://github.com/thewtex/Slicer/tree/minc-io-with-itk-update on Win10 with VS2013

**Comment by thewtex** [comment]
@lassoan Thanks for testing on the target compiler / platform.

It looks like the system has `rint`, but `HAVE_RINT` is not found -- this patch may address the issue:

  https://github.com/BIC-MNI/libminc/pull/83

**Comment by thewtex** [comment]
@lassoan Please try this branch:

  https://github.com/thewtex/Slicer/tree/minc-io-with-itk-update-2

**Comment by lassoan** [comment]
Build failed with this error:

35>  Project "C:\D\E\Slicer-0-build\ITKv4-build\ALL_BUILD.vcxproj" (1) is building "C:\D\E\Slicer-0-build\ITKv4-build\Modules\IO\MINC\src\ITKIOMINC.vcxproj" (110) on node 1 (default targets).
35>  Project "C:\D\E\Slicer-0-build\ITKv4-build\Modules\IO\MINC\src\ITKIOMINC.vcxproj" (110) is building "C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC\src\libminc\itkminc2.vcxproj" (111) on node 1 (default targets).
35>  InitializeBuildStatus:
35>    Touching "itkminc2.dir\Release\itkminc2.tlog\unsuccessfulbuild".
35>  CustomBuild:
35>    All outputs are up-to-date.
35>  ClCompile:
35>    C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\x86_amd64\CL.exe /c /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\ZLIB\src" /I"C:\D\E\Slicer-0-build\zlib-install\include" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\HDF5\src" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\HDF5\src" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\KWSys\src" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libcommon" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libsrc2" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\volume_io\Include" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\HDF5\src\itkhdf5" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\HDF5\src\itkhdf5" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\HDF5\src\itkhdf5\src" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC\src\libminc" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libsrc" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\ezminc" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\HDF5\src\itkhdf5\c++\src" /I"C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\HDF5\src\itkhdf5\c++\src" /I"C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\HDF5\src\itkhdf5\src" /nologo /W0 /WX- /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D _CRT_FAR_MAPPINGS_NO_DEPRECATE /D _CRT_IS_WCTYPE_NO_DEPRECATE /D _CRT_MANAGED_FP_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _CRT_SECURE_NO_DEPRECATE /D _CRT_SECURE_NO_DEPRECATE_GLOBALS /D _CRT_SETERRORMODE_BEEP_SLEEP_NO_DEPRECATE /D _CRT_TIME_FUNCTIONS_NO_DEPRECATE /D _CRT_VCCLRIT_NO_DEPRECATE /D _SCL_SECURE_NO_DEPRECATE /D HAVE_CONFIG_H /D MINC2=1 /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /GR /Fo"itkminc2.dir\Release\\" /Fd"itkminc2.dir\Release\vc120.pdb" /Gd /TC /errorReport:queue  /bigobj "C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libsrc2\dimension.c" "C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libsrc2\hyper.c"
35>    dimension.c
35>C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\math.h(515): error C2375: 'rint' : redefinition; different linkage [C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC\src\libminc\itkminc2.vcxproj]
35>            c:\d\e\slicer-0-build\itkv4\modules\thirdparty\minc\src\libminc\libsrc2\minc2_private.h(210) : see declaration of 'rint'
35>    hyper.c
35>C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libsrc2\hyper.c(52): error C2084: function 'double copysign(double,double)' already has a body [C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC\src\libminc\itkminc2.vcxproj]
35>            C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\math.h(483) : see previous definition of 'copysign'
35>C:\D\E\Slicer-0-build\ITKv4\Modules\ThirdParty\MINC\src\libminc\libsrc2\hyper.c(65): error C2084: function 'double round(double)' already has a body [C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC\src\libminc\itkminc2.vcxproj]
35>            C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\math.h(516) : see previous definition of 'round'
35>    Generating Code...
35>  Done Building Project "C:\D\E\Slicer-0-build\ITKv4-build\Modules\ThirdParty\MINC\src\libminc\itkminc2.vcxproj" (default targets) -- FAILED.
35>  Done Building Project "C:\D\E\Slicer-0-build\ITKv4-build\Modules\IO\MINC\src\ITKIOMINC.vcxproj" (default targets) -- FAILED.

I start a clean build now.

@thewtex Would you like to get access to the computer so that you can test and fix the errors directly? If yes then send me a message in private.

**Comment by thewtex** [comment]
@lassoan Thanks for checking.

If the error persists after a fresh build, I'll try to reproduce on a different box, and, if I still persists, debugging on the target will be very helpful.

**Comment by lassoan** [comment]
The error is the same when starting the build from scratch. See full build log here (first error is around line 79121): https://1drv.ms/t/s!Arm_AFxB9yqHr61zxIsdFj_PLLf5qw

**Comment by thewtex** [comment]
Rebased on Slicer master, which now includes updated ITK, which includes updated libminc. These include fixes for Visual Studio.

File: Libs/MRML/Core/vtkDataFileFormatHelper.cxx [hunk 1]
```diff
@@ -55,6 +55,7 @@ ITKImageFileFormat FileFormatTable[] =
   {"MGHImageIO", "Uncompressed FreeSurfer pixel data in binary", "MGH", ".mgh"},
   {"MGHImageIO", "Gzip Compressed FreeSurfer pixel data in binary", "MGH", ".mgz"},
   {"MGHImageIO", "Gzip Compressed FreeSurfer pixel data in binary", "MGH", ".mgh.gz"},
+  {"MINCImageIO", "Optional compression. Binary pixel data", "MINC", ".mnc"},
```
Comment (jcfr) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: Should a line line like this be added then:

```
{"MINCImageIO", "Gzip Compressed binary pixel data", "MINC", ".mnc.gz"},
```

Comment (gdevenyi) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: Should be careful here, classically, .mnc.gz files are "MINC1" aka NetCDF files, which the ITK MINC IO can't open.

.mnc files are now typically MINC2 files which are HDF5 with internal zlib compression built into the filespec.

Comment (vfonov) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: Integrated MINC ITK reader is based only on MINC2 library, and is not capable of reading MINC1 files. 
If one would compile full minc library (with MINC1 & MINC2 support) and incorporated it into Slicer, then ITK would be able to read .mnc.gz files.

Comment (jcfr) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: Thanks for the feedback. 

Would the following be correct then:

```
{"MINCImageIO", "MINC2 binary pixel data using HDF5 with internal zlib compression", "MINC2", ".mnc"},
```

Comment (vfonov) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: Yes, looks right.

Comment (vfonov) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: compression have to be enabled though, through SetUseCompression / UseCompressionOn - methods of ImageIOBase

Comment (jcfr) [inline Libs/MRML/Core/vtkDataFileFormatHelper.cxx]: Thanks for the feedback, I will plan on using this then:

```
{"MINCImageIO", "MINC2 binary pixel data using HDF5 with optional internal zlib compression", "MINC2", ".mnc"},
```

**Comment by jcfr** [comment]
Integrated in [r26011](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26011)

@lassoan That said, could you confirm that loading the sample data `pd_z-_float_xyz.mnc` gives the expected results:

http://slicer.kitware.com/midas3/item/286044

mnc_slicer

**Comment by lassoan** [comment]
I've tested this and Reading/writing .mnc files work well on Windows.

**Comment by thewtex** [comment]
:champagne: Thanks all for working together to make this happen!

**Comment by jcfr** [comment]
@thewtex when compiling Slicer using centos 5 with g++ 4.8.2 from the latest version of [docker-centos-build](https://github.com/thewtex/docker-centos-build), it reports the following error:

```
/usr/src/Slicer-build/ITKv4/Modules/ThirdParty/MINC/src/libminc/libsrc2/hyper.c:64:1: error: redefinition of ‘round’
 round(double x)
 ^
```

**Comment by jcfr** [comment]
The try_compile associated with [Modules/ThirdParty/MINC/src/libminc/CMakeLists.txt#L125](https://github.com/InsightSoftwareConsortium/ITK/blob/bd5667633802274e0cedf983d88c623006f92057/Modules/ThirdParty/MINC/src/libminc/CMakeLists.txt#L125) returns the following:

```
/opt/rh/devtoolset-2/root/usr/bin/gcc    -Wno-uninitialized -Wno-unused-parameter -Wall -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -funit-at-a-time -Wno-strict-overflow  -w   -fPIC  -fPIE -o CMakeFiles/cmTC_5d14a.dir/CheckSymbolExists.c.o   -c /usr/src/Slicer-build/ITKv4-build/CMakeFiles/CMakeTmp/CheckSymbolExists.c
/usr/src/Slicer-build/ITKv4-build/CMakeFiles/CMakeTmp/CheckSymbolExists.c: In function ‘main’:
/usr/src/Slicer-build/ITKv4-build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:19: error: ‘round’ undeclared (first use in this function)
   return ((int*)(&round))[argc];
                   ^
/usr/src/Slicer-build/ITKv4-build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:19: note: each undeclared identifier is reported only once for each function it appears in
```

**Comment by jcfr** [comment]
The issue is that the C99 extension is not enabled during introspection, adding the following to the CMakeLists addressed the problem (adapted from  [python-cmake-buildsystem/python-cmake-buildsystem](https://github.com/python-cmake-buildsystem/python-cmake-buildsystem)):

```
# Convenient macro allowing to update CMAKE_REQUIRED_DEFINITIONS
set(CMAKE_REQUIRED_DEFINITIONS )
macro(set_required_def var value)
  set(${var} ${value})
  list(APPEND CMAKE_REQUIRED_DEFINITIONS "-D${var}=${value}")
endmacro()

set_required_def(_GNU_SOURCE 1)       # Define on Linux to activate all library features
```

**Comment by jcfr** [comment]
The following commit should be backported and integrated: https://github.com/Slicer/ITK/commit/3cc1a24d456ba41abff3e8508765f8f8d71bfcab

**Pull request closed without merging**